### ISSUE TITLE
Avoid rerunning id_tap and id_print with jax.remat

### ISSUE THREAD

**Issue by sbodenstein** [google/jax#5886]
One of the biggest annoyances with using the fantastic `host_callback.id_tap` as a logging solution is its behaviour in a `jax.remat` (making a call to `id_tap` for every `remat`). When you have a codebase with multiple layers of nested `remat`s, you get spammed with calls to `host_callback.id_tap`, which both increases the chances of a slowdown due to Python callback overhead, but also makes it hard to efficiently get rid of spurious duplicated logs coming from `remat`s.

Is there any way of implementing an option to `id_tap` to avoid calling it multiple times in a `remat`, or some other workaround?

@mattjj , @gnecula

**Comment by gnecula** [comment]
What is special about `id_tap` is that it is effectful. The rest of JAX is functional, and transformations like `remat` have no effect on the semantics. 

We are thinking about adding effects to the JAX type system, and with those we may be able to implement some other semantics for efectful operations. What would be a specification of how effectful operations should behave in presence of `remat`? 

With the current implementation `id_tap` is assumed to be idempotent, like any other JAX primitive, and it places the burden on the user to handle duplicate output. If we agree on a semantics then I can think how to implement it.

**Comment by sbodenstein** [comment]
> What would be a specification of how effectful operations should behave in presence of remat?

Remat is tricky. I honestly don't know what the solution is here. You really want an effectful op to know whether its being rerun as part of a remat, but I'm not sure if this would be possible to provide.

Would custom interpreters allow one to remove repeated calls to something like `id_tap` in a remat? @mattjj

**Comment by mattjj** [comment]
With custom interpreters, anything is possible!

I agree with George that the right idea here is to model `id_tap`'s effectfulness explicitly throughout the system. Right now much of the system assumes purity (like `remat`, and `vmap`-of-`cond`), but we have a few impure operations, and so we should at least not handle them incorrectly.

**Comment by sbodenstein** [comment]
So I think I have a workaround for this issue I'm happy with for logging (unless I've missed something where this fails):
```
import jax
import jax.numpy as jnp
from jax.experimental import host_callback

def safe_id_tap(func, x):
  @jax.custom_vjp
  def _safe_id_tap(x1):
    return host_callback.id_tap(func, x1)

  def _safe_id_tap_fwd(x):
    return x, (x)

  def _safe_id_tap_bwd(res, g):
    g = host_callback.id_tap(func, res, result = g)
    return (g,)

  _safe_id_tap.defvjp(_safe_id_tap_fwd, _safe_id_tap_bwd)
  return _safe_id_tap(x)

# @jax.remat
def f(x):
  x = jnp.sin(x)
  y = jnp.ones(2)
  y = safe_id_tap(print, y)
  x = safe_id_tap(print, x)
  x = jnp.sum(jnp.cos(x) * y)
  return x

a=jax.jit(jax.grad(f))(1.2)

# [1. 1.] ()
# 0.9320391 ()
```
This also provides a way to avoid the information-free gradient callbacks issue (https://github.com/google/jax/issues/5883) that produces 2x the number of Python callbacks during `jax.grad`.
